realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java

```diff
@@ -317,7 +319,33 @@ public void insertToRealm_NullPrimaryKey() {
     }
 
     @Test
-    public void insertOrUpdateToRealm() {
+    public void insert_duplicatedPrimaryKeyFails() {
+
+        // Single object with 2 references to two objects with the same ID
+        AllJavaTypes obj = new AllJavaTypes(2);
+        obj.setFieldList(new RealmList<AllJavaTypes>(new AllJavaTypes(1), new AllJavaTypes(1)));
+        realm.beginTransaction();
+        try {
+            realm.insert(obj);
+            fail();
+        } catch (RealmPrimaryKeyConstraintException ignored) {
```

Comment:
Note. It would work, if the list referenced the same Java object, e.g. `new RealmList<>(obj, obj)`. Then we correctly detect that it is the same object, but if you have to distinct java objects we can no longer assume anything